examples/bigquery_user_info_updater_tool/README.md

```diff
@@ -0,0 +1,317 @@
+--updates_table_id=<ID of the table that will hold updates (i.e. user_info_updates)> \
+--temp_updates_table_id=<ID of the intermediary table (i.e. temp_user_info_updates)>
+```
+ 
+* Configure your Dataflow pipeline to start writing updates to the `user_info_updates` table. 
+
+#### Updating User Info
+
+In order to update user info, run the [update_user_info.py](bigquery_user_info_updater/update_user_info.py) script using the 
+following command:  
+```
+python biguqery_user_info_updater/update_user_info.py \
+--project_id=<ID of your project> \
+--schema_path=<Path to the JSON user schema described in the first step> \
+--dataset_id=<ID of the dataset that will hold the user tables> \
+--final_table_id=<ID of the final golden record table (i.e. user_info_final)> \
+--updates_table_id=<ID of the table that will hold updates (i.e. user_info_updates)> \
+--temp_updates_table_id=<ID of the intermediary table (i.e. temp_user_info_updates)> \
+--user_id_field_name=<Name of the field that identifies unique users (i.e. userId) \
+--ingest_timestamp_field_name=<Name of the timestamp field that marks the ingestion of user rows (i.e. ingestTimestamp)> 
+
+```
+
+#### Scheduling Updates with CronJobs on GKE
+Once the script and queries have been setup and tested, the next step (if desired) is to schedule the job to run so that user info can be periodically updated.  There are few options for scheduling using different technologies (ie GKE, Cloud Composer,..) 
+For the purpose of this article we decided to leverage [GKE Cronjobs](https://cloud.google.com/kubernetes-engine/docs/how-to/cronjobs) (currently in Beta). We will take
+advantage of the `concurrencyPolicy` setting available in CronJobs on GKE. We will set `concurrencyPolicy` to `Forbid` to prevent a job 
+from running if a previous job has not finished. If a job kicks off while another is running, the data in the final table
+will be at risk for inaccuracies and duplicates.
+
+##### Prepare Files
+In order to schedule the script via a CronJob in GKE, several files will need to be prepared and created. 
+
+1. First make sure to `cd` into the [`cron`](cron) directory of the this repo. Once there, copy the `bigquery_user_info_updater` directory into the cron directory so that the cron files will have access to the script: 
+```cp -r ../../bigquery_user_info_updater/ bigquery_user_info_updater```
+
+2. Uncomment the commented lines in [`cron/bigquery_user_updater.sh`](cron/bigquery_user_updater.sh) with your own values.
+
+3. Uncomment the commented lines in [`cron/Dockerfile`](cron/Dockerfile) with your own values. While you can set the schedule and most name parameters to whatever you 
+would like, the important thing is that you absolutely set the `concurrencyPolicy` to `Forbid` to prevent inaccuracies and duplicates. Also, it is important that you keep `env.name` set to `GOOGLE_APPLICATION_CREDENTIALS`. This will allow your container to export your google 
```

Comment:
jaketf on Aug 27, 2019 Member
Nit: This important thing is really important I would mention this above in "Scheduling Updates with CronJobs on GKE"

 @jaketf
jaketf on Aug 27, 2019 Member
Also GKE CronJobs are an at least once guarantee.

While this may be rare you might encounter it if you are running jobs every minute.
If you have a CronJob get evicted and scheduled on another pod while the BQ MERGE query is in flight you will not get the intended behavior if you fire another copy of the sameMERGE query.

A system going to production should keep the state of what time interval / updates have been claimed (and are thus in flight) or completed and your cron job should not resubmit the same merge query but instead log a warning and wait for the in-flight job to finish and do any post processing (like marking it complete in the tracking system).

 @jaketf
jaketf on Aug 27, 2019 Member
@ryanmcdowell another example of a place we have not-quite idempotent cron jobs that need a central place to manage state for claimed / completed time intervals.

 @annarudy
annarudy on Oct 4, 2019 Author Member
@jaketf in the original solution, I had a BQ table with just one of job data, including a 'running' boolean field to indicate if a job was currently running. Is that the type of solution you were getting at? It felt a little strange to have such a small BQ table for this purpose, but we couldn't use GKE CronJob features like concurrencyPolicy since we were using traditional cron jobs.

 @jaketf
jaketf on Oct 4, 2019 Member
Yeah that sounds close to what I'm suggesting. Agreed, BQ is not a natural place for something like that.
Could you share the schema for that table and how your cronjob handled if a job already existed for the time range?

 @annarudy
annarudy on Oct 14, 2019 Author Member
sure here is the schema: [ { "name": "latest_merge", "type": "TIMESTAMP", "mode": "REQUIRED" }, { "name": "running", "type": "BOOLEAN", "mode": "REQUIRED" } ]

 @annarudy
annarudy on Oct 14, 2019 Author Member
Basically, each time the cron job kicked off, the python script ran a query on the metadata table, and if running was set to TRUE, then the job would stop and a message saying that the job is already run would be logged out. If it was set to FALSE, then the script would set it to TRUE and proceed with the job.

 @jaketf
jaketf on Oct 17, 2019 Member
I think a mutex table like that is very important to a solution like this.

However, your solution is missing a few things to be robust:

There needs to be a way to release an abandoned claim. Consider the scenario where a cron job claims the time interval and then pod gets evicted (before actually starting the BQ job). Your next cron job will do nothing because your metadata table thinks there's a BQ job is running. Similarly if there's a transient issue that causes the BQ job to fail (think throttling DML statements or BQ load jobs based on quota/limits) you may end up with a claimed time interval that isn't actually in flight. This could be some other GKE cron job / agent to release abandoned claims with the logic: "Hey Mutex table says my job is running but there are no BQ jobs running from this service account so release the claim".
Agreed a small BQ table like this seems silly and the latency of querying it to get a simple flag unnecessary. This is basically an application state store. Cloud SQL, Redis or firestore would be better for this use case. They all have trade-offs I think it'd be simplest to use Cloud SQL (run sql proxy sidecar in GKE) most performant to run redis operator on the same gke cluster and most "cloud native" to use firestore.
Your Mutex solution should be able to handle the following:

Catchup by claiming multiple time intervals : What if your a cron job fails for any reason. Should the next cron job only fire a query to process a single time interval or should it attempt to process all three failed time intervals? (hint: if it only processes a single time interval your system never catches up.). You can either achieve this by having a single job claim multiple intervals or have each row have a interval_start interval_end kind of field.
Who claimed this time interval?: for debugging it could be helpful for the mutex system to keep track of the the bq job id responsible for each time interval. This would be helpful on pod eviction for reattaching to an existing bq job and polling til it succeeds.
More enumerated statuses that running / not running: PENDING RUNNING SUCCESS FAILED . This can allow your cron job to update the mutex table in the event of a failure and indicate that the next cron job should retry (or not if it is an unretriable error e.g. schema mismatch). Any new cron job should claim all PENDING time intervals and process them in a single query job if possible). It should also check if a time interval has been marked as running for way longer than expected, check if it's BQ jobs status and mark it as PENDING to retry or FAILED and send some alert to stackdriver because something went really wrong.
Reprocessing: you should be able to manually mark many time intervals as pending to reprocess them.

Reply:
given that the scheduled interval is 5 min (I was using 1 min for testing), and given the amount of time I have left, I think it's best to not add the extra complexity to this pr for now. Instead I could add a warning message. But, I really appreciate you taking the time to think all of this out.